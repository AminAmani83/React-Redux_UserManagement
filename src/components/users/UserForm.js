import React from "react";
import TextInput from "../common/TextInput";
import SelectInput from "../common/SelectInput";
import PropTypes from "prop-types";

const UserForm = ({
  handleSubmit,
  handleFieldChange,
  validationErrors,
  user,
  addresses,
  isSaving,
  isLoading,
  preloadFields,
}) => {
  return (
    <form onSubmit={handleSubmit}>
      <div className=" user-form">
        <div className="form-group">
          <TextInput
            label="Name:"
            name="name"
            value={preloadFields && isLoading ? "Loading..." : user.name}
            disabled={isSaving}
            handleFieldChange={handleFieldChange}
          />
          <div
            className={
              validationErrors.name
                ? "alert alert-danger input-field-error"
                : ""
            }
          >
            {validationErrors.name}
          </div>
        </div>

        <div className="form-group">
          <TextInput
            label="Age:"
            name="age"
            value={preloadFields && isLoading ? "Loading..." : user.age}
            disabled={isSaving}
            handleFieldChange={handleFieldChange}
          />
          <div
            className={
              validationErrors.age ? "alert alert-danger input-field-error" : ""
            }
          >
            {validationErrors.age}
          </div>
        </div>

        <div className="form-group">
          <SelectInput
            label="Address:"
            name="addressId"
            addressId={user.addressId}
            options={addresses.map((a) => ({
              value: a.id,
              text: `${a.province}, ${a.city}, ${a.street}`,
            }))}
            defaultSelectedText={
              addresses.length > 0 ? "Select an Address" : "Loading..."
            }
            disabled={isSaving}
            handleFieldChange={handleFieldChange}
          />
          <div
            className={
              validationErrors.addressId
                ? "alert alert-danger input-field-error"
                : ""
            }
          >
            {validationErrors.addressId}
          </div>
        </div>

        <button
          type="submit"
          className="btn btn-info"
          disabled={isSaving}
          style={{ width: 100 }}
        >
          {isSaving ? "Saving..." : "Save"}
        </button>
      </div>
    </form>
  );
};

UserForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  handleFieldChange: PropTypes.func.isRequired,
  user: PropTypes.object.isRequired,
  addresses: PropTypes.arrayOf(PropTypes.object).isRequired,
  isSaving: PropTypes.bool.isRequired,
  validationErrors: PropTypes.object.isRequired,
  isLoading: PropTypes.bool.isRequired,
  preloadFields: PropTypes.bool.isRequired,
};

export default UserForm;
