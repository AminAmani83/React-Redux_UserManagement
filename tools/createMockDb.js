/* eslint-disable no-console */
const fs = require("fs");
const path = require("path");
const mockData = require("./mockData");

const { react_users, react_addresses } = mockData;
const data = JSON.stringify({ react_users, react_addresses });
const filepath = path.join(__dirname, "db.json");

fs.writeFile(filepath, data, function (err) {
  err ? console.log(err) : console.log("Mock DB created.");
});
