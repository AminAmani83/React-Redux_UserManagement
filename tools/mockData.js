const react_users = [
  {
    id: "e452b8a9-976d-42e3-a70d-3b3e6a7d53c0",
    name: "Amin1",
    age: 23,
    addressId: "52a652d5-8667-4b99-9124-350165f1dbc0",
  },
  {
    id: "e452b8a9-976d-42e3-a70d-3b3e6a7d53c1",
    name: "Amin2",
    age: 21,
    addressId: "52a652d5-8667-4b99-9124-350165f1dbc1",
  },
  {
    id: "e452b8a9-976d-42e3-a70d-3b3e6a7d53c2",
    name: "Amin4",
    age: 26,
    addressId: "52a652d5-8667-4b99-9124-350165f1dbc0",
  },
  {
    id: "e452b8a9-976d-42e3-a70d-3b3e6a7d53c3",
    name: "Amin3",
    age: 20,
    addressId: "52a652d5-8667-4b99-9124-350165f1dbc0",
  },
  {
    id: "e452b8a9-976d-42e3-a70d-3b3e6a7d53c4",
    name: "Amin6",
    age: 25,
    addressId: "52a652d5-8667-4b99-9124-350165f1dbc2",
  },
  {
    id: "e452b8a9-976d-42e3-a70d-3b3e6a7d53c5",
    name: "Amin7",
    age: 27,
    addressId: "52a652d5-8667-4b99-9124-350165f1dbc1",
  },
  {
    id: "e452b8a9-976d-42e3-a70d-3b3e6a7d53c6",
    name: "Amin5",
    age: 26,
    addressId: "52a652d5-8667-4b99-9124-350165f1dbc2",
  },
  {
    id: "e452b8a9-976d-42e3-a70d-3b3e6a7d53c7",
    name: "Amin8",
    age: 28,
    addressId: "52a652d5-8667-4b99-9124-350165f1dbc1",
  },
  {
    id: "e452b8a9-976d-42e3-a70d-3b3e6a7d53c8",
    name: "Amin9",
    age: 29,
    addressId: "52a652d5-8667-4b99-9124-350165f1dbc0",
  },
  {
    id: "e452b8a9-976d-42e3-a70d-3b3e6a7d53c9",
    name: "Amin10",
    age: 30,
    addressId: "52a652d5-8667-4b99-9124-350165f1dbc2",
  },
];

const react_addresses = [
  {
    id: "52a652d5-8667-4b99-9124-350165f1dbc0",
    province: "QC",
    city: "Montreal",
    street: "101 Sherbrooke West",
  },
  {
    id: "52a652d5-8667-4b99-9124-350165f1dbc1",
    province: "QC",
    city: "Montreal",
    street: "202 Sherbrooke West",
  },
  {
    id: "52a652d5-8667-4b99-9124-350165f1dbc2",
    province: "QC",
    city: "Montreal",
    street: "303 Sherbrooke West",
  },
];

const newUser = {
  id: "",
  name: "",
  age: null,
  addressId: "",
};

const newAddress = {
  id: "",
  province: "",
  city: "",
  street: "",
};

// Using CommonJS style export so we can consume via Node (without using Babel-node)
module.exports = {
  newUser,
  newAddress,
  react_users,
  react_addresses,
};
