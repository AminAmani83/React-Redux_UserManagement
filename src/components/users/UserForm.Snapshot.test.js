import React from "react";
import UserForm from "./UserForm";
import renderer from "react-test-renderer";
import { users, addresses } from "../../../tools/mockData";

it("changes the submit button label to 'Saving...' when saving is true", () => {
  const tree = renderer.create(
    <UserForm
      handleSubmit={jest.fn()}
      handleFieldChange={jest.fn()}
      user={users[0]}
      validationErrors={{}}
      addresses={addresses}
      isSaving={true}
      isLoading={false}
    />
  );
  expect(tree).toMatchSnapshot();
});

it("changes the submit button label to 'Save' when saving is false", () => {
  const tree = renderer.create(
    <UserForm
      handleSubmit={jest.fn()}
      handleFieldChange={jest.fn()}
      user={users[0]}
      validationErrors={{}}
      addresses={addresses}
      isSaving={false}
      isLoading={false}
    />
  );
  expect(tree).toMatchSnapshot();
});
