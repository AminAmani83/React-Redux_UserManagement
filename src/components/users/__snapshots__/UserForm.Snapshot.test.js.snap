// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes the submit button label to 'Save' when saving is false 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <div
    className="form-group"
  >
    <label
      htmlFor="name"
    >
      Name:
    </label>
    <input
      className="form-control"
      disabled={false}
      id="name"
      name="name"
      onChange={[MockFunction]}
      type="text"
      value="Amin1"
    />
    <div
      className=" alert input-field-error"
    />
    <label
      htmlFor="age"
    >
      Age:
    </label>
    <input
      className="form-control"
      disabled={false}
      id="age"
      name="age"
      onChange={[MockFunction]}
      type="text"
      value={21}
    />
    <div
      className=" alert input-field-error"
    />
    <label
      htmlFor="addressId"
    >
      Address:
    </label>
    <select
      className="form-control"
      disabled={false}
      id="addressId"
      name="addressId"
      onChange={[MockFunction]}
      value={1}
    >
      <option
        value={null}
      >
        Select an Address
      </option>
      <option
        value={1}
      >
        QC, Montreal, 101 Sherbrooke West
      </option>
      <option
        value={2}
      >
        QC, Montreal, 202 Sherbrooke West
      </option>
      <option
        value={3}
      >
        QC, Montreal, 303 Sherbrooke West
      </option>
    </select>
    <div
      className=" alert input-field-error"
    />
    <button
      className="btn btn-info"
      disabled={false}
      style={
        Object {
          "width": 100,
        }
      }
      type="submit"
    >
      Save
    </button>
  </div>
</form>
`;

exports[`changes the submit button label to 'Saving...' when saving is true 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <div
    className="form-group"
  >
    <label
      htmlFor="name"
    >
      Name:
    </label>
    <input
      className="form-control"
      disabled={true}
      id="name"
      name="name"
      onChange={[MockFunction]}
      type="text"
      value="Amin1"
    />
    <div
      className=" alert input-field-error"
    />
    <label
      htmlFor="age"
    >
      Age:
    </label>
    <input
      className="form-control"
      disabled={true}
      id="age"
      name="age"
      onChange={[MockFunction]}
      type="text"
      value={21}
    />
    <div
      className=" alert input-field-error"
    />
    <label
      htmlFor="addressId"
    >
      Address:
    </label>
    <select
      className="form-control"
      disabled={true}
      id="addressId"
      name="addressId"
      onChange={[MockFunction]}
      value={1}
    >
      <option
        value={null}
      >
        Select an Address
      </option>
      <option
        value={1}
      >
        QC, Montreal, 101 Sherbrooke West
      </option>
      <option
        value={2}
      >
        QC, Montreal, 202 Sherbrooke West
      </option>
      <option
        value={3}
      >
        QC, Montreal, 303 Sherbrooke West
      </option>
    </select>
    <div
      className=" alert input-field-error"
    />
    <button
      className="btn btn-info"
      disabled={true}
      style={
        Object {
          "width": 100,
        }
      }
      type="submit"
    >
      Saving...
    </button>
  </div>
</form>
`;
